packages/telemetry/src/managers/tracing.manager.ts

```diff
@@ -0,0 +1,152 @@
+        this.trace = new Trace();
+        this.trace.id = traceId ?? uuidv4();
+        this.trace.context = context;
+
+        const span = new Span(this);
+        span.id = uuidv4();
+        span.keyname = spanRootKeyname
+        span.trace = this.trace;
+        span.context = context;
+
+        this.trace.rootSpan = span;
+
+        this.spans[span.keyname] = span;
+    }
+
+    /**
+     * This method starts a new span.
+     *
+     * @param keyname
+     * @param parentId
+     * @param context
+     */
+    public startSpan(keyname: string, parentId?: string, context?: any) {
+        // Check if there's an active trace. If not, start one.
+        if(this.trace === undefined) {
+            this.startTracing(SpanKeynameEnum.RootExecution, undefined, context);
+        }
+
+        let parentSpan: Span = this.trace!.rootSpan;
+
+        // Check to find the parentId in our internal map of spans
+        if(parentId) {
+            parentSpan = this.spans[parentId] ?? parentSpan;
+        }
+
+        // Create the new span
+        const span = new Span(this);
+        span.id = uuidv4();
+        span.keyname = keyname
+        span.trace = this.trace!;
+        span.context = context;
+        span.parentSpan = parentSpan;
+        parentSpan.childSpans.push(span);
+
+        // Add it to the map of spans
+        this.spans[span.keyname] = span;
+
+        // Notify the Tracers that a new span was started.
+        this.tracers.forEach( (tracer:TracerInterface) => {
+            tracer.spanStartedStream?.push(span);
+        })
+    }
+
+    /**
+     * This method ends the span using a keyname.
+     *
+     * @param keyname
+     */
+    public endSpanKeyname(keyname: string) {
+        if(this.spans.hasOwnProperty(keyname) === false) {
+            return;
+        }
+
+        this.endSpan(this.spans[keyname]);
+    }
+
+    /**
+     * This methods ends the span by setting the end date and by calling the tracers.
+     *
+     * It will also end the trace if the rootspan is being ended.
+     *
+     * @param span
+     */
+    public endSpan(span: Span) {
+        if(span.endDate !== undefined) {
+            return;
+        }
+
+        span.endDate = Date.now();
+
+        // Notify the TraceListeners that the span was ended.
+        this.tracers.forEach( (tracer:TracerInterface) => {
+            tracer.spanEndedStream?.push(span);
+        })
+
+        // If the span is the root span, the trace has ended
+        if(span.keyname === this.trace?.rootSpan.keyname) {
+            this.endTrace()
+        }
+    }
+
+    /**
+     * This method ends the trace entirely.
+     */
+    public endTrace() {
+        if(this.trace === undefined || this.trace.hasEnded) {
+            return;
+        }
+
+        // End the trace.
+        this.trace.hasEnded = true;
+
+        // Loop over all the spans and end them if they are not already ended.
+        Object.keys(this.spans).forEach(spanKeyname => {
```

Comment:
Quand on fait Object.keys on a pas besoin de verifier hasOwnProperty ?

Reply:
Je ne pense pas non. c'est vraiment le for in messemble qu'on doit vérifier.